beangulp/testing.py

```diff
@@ -76,90 +122,72 @@ def log(msg, level=0, **kwargs):
 @click.option('--quiet', '-q', count=True, help="Suppress all output.")
 @click.option('--exitfirst', '-x', is_flag=True, help="Stop at the first test failure.")
 @click.pass_obj
-def _test(ctx, documents, expected, generate, verbose, quiet, exitfirst):
+def _test(ctx, documents: List[str], generate: bool,
+          verbose: bool, quiet: bool, exitfirst: bool):
     """Run the importer on all DOCUMENTS and verify that it produces the
     desired output.  The desired output is stored in Beancount ledger
     files with an header containing additional metadata.  These files
     are named as the SHA1 digest of the documents and can be generated
     with the --generate option.
 
-    DOCUMENTS can be files or directories.  Directories are walked and
-    the importer is called on all files.  When no DOCUMENTS path is
-    specified on the command line, the tests/documents/ directory
-    relative to the test script location is used.  All documents for
-    which an expected output file exist must be positively identified
-    by the importer.
-
-    Expected output files are searched in the tests/expected/$NAME/
-    directory relative to the test script location, where $NAME is
-    replaced with the importer name, if the path exists, or in the
-    tests/expected/ directory.  This path can be overwritten through
-    the --expected DIR option.
-
+    DOCUMENTS can be files or directories. Directories are walked and the
+    importer is called on all files found, except for .beancount files.
     """
     assert len(ctx.importers) == 1
     importer = ctx.importers[0]
 
-    import __main__
-    root = os.path.dirname(__main__.__file__)
-
-    if not documents:
-        documents = (os.path.join(root, 'tests','documents'), )
-
-    if not expected:
-        expected = os.path.join(root, 'tests', 'expected', importer.name())
-        if not os.path.exists(expected):
-            expected = os.path.join(root, 'tests', 'expected')
-
     verbosity = verbose - quiet
     log = logger(verbosity)
     failures = 0
 
-    for doc in walk(documents):
-        # unless verbose mode is enabled, do not output a newline so
+    for doc in walk(documents, r".*\.beancount$"):
+        # Unless verbose mode is enabled, do not output a newline so
         # the test result is printed on the same line as the test
         # document filename.
-        log(f'∙ {os.path.relpath(doc):}', nl=verbosity > 0)
-
-        # compute the path to the expected output file
-        path = os.path.join(expected, sha1sum(doc) + '.beancount')
-        f = cache.get_file(os.path.abspath(doc))
-
-        if importer.identify(f):
-            account, date, name, entries = run_importer(importer, f)
-            log(f'  {os.path.relpath(path):}', 1)
-            log(f'  {account.replace(":", "/"):}/{date:%Y-%m-%d}-{name:}', 1)
+        log(f'* {doc}', nl=verbosity > 0)
+
+        # Compute the path to the expected output file.
+        expected_filename = "{}.beancount".format(doc)
```

Comment:
The reason for using the SHA1 is to avoid ambiguity between changes in the test document and changes to the importer code. The SHA1 can be added to the expected file header. Another reason for using an unintelligible file name is to convey the idea that the user is not expected to touch it.

Reply:
For the first point: how does it avoid ambiguity? I don't understand. Adding the SHA1 to the expected file header would seem sensible, but this means that every change to the importers - even cosmetic changes - would require regenerating all the expected files. Not great.

For the second point, having an unintelligible filename just makes it harder to find the expected output. Plus users of this know the purpose of these files, I don't think it indicates anything useful. It's real nice to have the expected file right next to the downloaded file, right in the same directory.

Reply:
> For the first point: how does it avoid ambiguity? I don't understand. Adding the SHA1 to the expected file header would seem sensible, but this means that every change to the importers - even cosmetic changes - would require regenerating all the expected files. Not great.

The SHA1 of the test document, not the SHA1 of the importer implementation.

Reply:
So it's not about hashing the importer source... I guess I don't understand what problem using the SHA1 of the input file solves?
We need to be able to easily find and associate that file with its corresponding input.

Reply:
The idea is to protect against accidental changes or overwrites of the test documents (which are even more likely if the recommended test file location is not a dedicated directory).

Reply:
Oh I see. These should always be under source control, so I'm not too worried. Perhaps we can rename the files before they get overwritten by default, and have an option to disable that to the new generate command.

Reply:
I was more thinking of someone obliterating by mistake his test document, not the expected output file. However, if we stick to the separate command for generating the expected output file, it makes sense to refuse overwriting unless a `--force` option is provided. I didn't do that in my implementation because the CLI wound have been awkward.